### ISSUE TITLE
[Feature Request] ImplicitLambdaParameter rules

### ISSUE THREAD

**Issue by QuinnBast** [detekt/detekt#4131]
Is there a rule that forces lambda parameters to be defined? My team would love to enforce something like this and completely require all lambda variables to be named.

```
var.map { it.value } // BAD!
var.map { namedVar -> namedVar.value } // Force a name!
```

I am also looking for a rule for lambda variable length requirements.

```
var.map { a -> a.value } // BAD!
var.map { moreSpecificName -> moreSpecificName.value } // Good.
```

_Originally posted by @QuinnBast in https://github.com/detekt/detekt/issues/3245#issuecomment-926899323_

**Comment by BraisGabin** [comment]
I think that we could change `MultilineLambdaItParameter` to fix the first point. But I think that it would be a Breaking change because we should rename it to `ImplicitItParameter`. And then have a flag (on by default) to "ignore one line lambdas".

I think that the second point should be handled by [VariableMinLength](https://detekt.github.io/detekt/naming.html#variableminlength). That or we should create one for `ParameterMinLength`.

**Comment by QuinnBast** [comment]
An alternative is to use the `VariableNaming` rule within lambdas. It would be possible to send it the following rule to error on variables named `it`:

```yaml
  VariableNaming:
    active: true
    excludes: "**/test/**,**/androidTest/**,**/*.Test.kt,**/*.Spec.kt,**/*.Spek.kt"
    variablePattern: '\b(?!it\b)([a-z][A-Za-z0-9]*)' # Error on variables named "it"
    privateVariablePattern: '\b(?!it\b)((_)?[a-z][A-Za-z0-9]*)' # Error on variables named "it"
    ignoreOverridden: true
```

This would make any variable named `it` fail the `VariableNaming` check. We just need to apply it to lambdas.

**Comment by chao2zhang** [comment]
`VariableNaming` and `VariableMinLength` checking class properties and variables. If it also checks lambda expression then we are probably adding too many responsibilities in the rule.

I would vote for new rules.

**Comment by cortinico** [comment]
> I would vote for new rules.

+1 specifically for the part related to:

> I am also looking for a rule for lambda variable length requirements.

**Comment by BraisGabin** [comment]
I Implemented the rule `LambdaParameterNaming` at #4147 because we wasn't checking those parameter names.

Do we want to have `LambdaParameterMinLenght` and `LambdaParameterMaxLenght`? I don't think so. The community that wants that I think is small and you can archive it with regex: `[a-z][A-Za-z0-9]{1,63}` (valid if size >= 2 and <= 64). This would fix the second point from the first post.

What should we do about the first point?

**Comment by chao2zhang** [comment]
Actually, `LambdaParameterNaming` in `https://github.com/detekt/detekt/pull/4147` completely covers the existing rule for `ExplicitItLambdaParameter`. For the first point, we should probably deprecate `ExplicitItLambdaParameter` and create a new rule with a name `RequireNamedLambdaParameter`

**Comment by marschwar** [comment]
I disagree. Both rules cover different use cases. `ExplicitItLambdaParameter` only verifies that the parameter is not named `it` and `LambdaParameterNaming` checks for certain patterns.

**Comment by BraisGabin** [comment]
My thoughts:

- We agree that we want a rule called `RequiredNamedLambdaParameter`. Probably I would not enable it in my project but it seems like a good rule to have in the default set.
- If we implement `RequiredNamedLambdaParameter` we can get rid of `MultilineLambdaItParameter`. We can replace it with a flag in `RequiredNamedLambdaParameter`. Something like `ignoreOneLineLambda` or similar.
- `LambdaParameterNaming` could replace `ExplicitItLambdaParameter` with this regex: `(?!it$)[a-z][A-Za-z0-9]*|_` (or something like this). But even if it is possible I think that this rule is good to avoid "cheating" `RequiredNamedLambdaParameter`.

**Comment by marschwar** [comment]
> We agree that we want a rule called RequiredNamedLambdaParameter. Probably I would not enable it in my project but it seems like a good rule to have in the default set.

Agreed.

> If we implement RequiredNamedLambdaParameter we can get rid of MultilineLambdaItParameter. We can replace it with a flag in RequiredNamedLambdaParameter. Something like ignoreOneLineLambda or similar.

Good idea. I would actually set this to true by default.
 
> LambdaParameterNaming could replace ExplicitItLambdaParameter with this regex: (?!it$)[a-z][A-Za-z0-9]*|_ (or something like this). But even if it is possible I think that this rule is good to avoid "cheating" RequiredNamedLambdaParameter.

I am against combining those two rules by changing the regular expression. Either we keep both rules or we no longer prohibit `it`as a lambda parameter name.

**Comment by BraisGabin** [comment]
> I am against combining those two rules by changing the regular expression. Either we keep both rules or we no longer prohibit itas a lambda parameter name.

My vote goes to keep both. Even we *can* merge them, it doesn't mean that we should do it.

**Comment by QuinnBast** [comment]
Keeping the existing rule and just adding a new one makes backwards compatibility easier :)
